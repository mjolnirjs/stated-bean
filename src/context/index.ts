export * from './StatedBeanConsumer';
export * from './StatedBeanContext';
export * from './StatedBeanProvider';
