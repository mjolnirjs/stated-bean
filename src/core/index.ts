export * from './StatedBeanApplication';
export * from './StatedBeanContainer';
export * from './StatedBeanFactory';
export * from './Symbols';
export * from './AutoBind';
export * from './EffectEvent';
