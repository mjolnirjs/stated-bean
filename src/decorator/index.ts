export * from './Effect';
export * from './ObservableProps';
export * from './AfterProvided';
export * from './Props';
export * from './Stated';
export * from './StatedBean';
