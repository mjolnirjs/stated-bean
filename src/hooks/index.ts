export * from './useBean';
export * from './useContainer';
export * from './useInject';
export * from './useObservable';
export * from './useObserveEffect';
