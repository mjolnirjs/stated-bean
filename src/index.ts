export * from './core';
export * from './decorator';
export * from './hooks';
export * from './context';
export * from './types';
