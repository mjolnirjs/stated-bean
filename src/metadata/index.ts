import { StatedBeanMetaStorage } from './storage';

export function getMetadataStorage(): StatedBeanMetaStorage {
  return (
    (window as any).StateBeanMetadataStorage ||
    ((window as any).StateBeanMetadataStorage = new StatedBeanMetaStorage())
  );
}

export * from './storage';
