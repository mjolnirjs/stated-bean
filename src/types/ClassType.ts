// eslint-disable-next-line @typescript-eslint/no-type-alias
export type ClassType<T = unknown> = new (...args: unknown[]) => T;

// eslint-disable-next-line @typescript-eslint/no-type-alias
export type InstanceType<T extends ClassType> = T extends ClassType<infer R>
  ? R
  : never;

// eslint-disable-next-line @typescript-eslint/no-type-alias
export type FunctionPropertyNames<T> = {
  [K in keyof T]: T[K] extends Function ? K : never;
}[keyof T];
