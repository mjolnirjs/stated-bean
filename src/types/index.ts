export * from './ClassType';
export * from './StatedMetadata';
export * from './Provider';
export * from './StatedBeanType';
export * from './EffectEventValue';
