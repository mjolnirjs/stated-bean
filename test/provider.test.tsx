import { getMetadataStorage } from '../src/metadata';
import {
  StatedBean,
  Stated,
  StatedBeanConsumer,
  StatedBeanProvider,
  useStatedBean,
  StatedBeanContextValue,
} from '../src';
import { ClassType } from '../src/types/ClassType';

import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

describe('react provider', () => {
  let TestStatedBean: ClassType;

  class T {
    t!: number;
  }

  beforeAll(() => {
    getMetadataStorage().clear();

    @StatedBean()
    class SampleStatedBean {
      @Stated()
      statedField: number;

      @Stated()
      statedField2: string;

      constructor() {
        this.statedField = 0;
        this.statedField2 = 'testStatedField';
      }

      addStatedField = () => {
        this.statedField += 1;
      };
    }

    TestStatedBean = SampleStatedBean;
  });

  it('StatedBeanProvider', () => {
    const Sample = () => {
      const bean = useStatedBean(TestStatedBean);

      // eslint-disable-next-line react-hooks/rules-of-hooks
      expect(() => useStatedBean(T)).toThrow();

      expect(bean).not.toBeNull();
      expect(bean.statedField).toEqual(0);

      return (
        <div>
          {bean.statedField}{' '}
          <button id="addBtn" onClick={bean.addStatedField}>
            add
          </button>
        </div>
      );
    };

    const App = () => {
      return (
        <StatedBeanProvider types={[TestStatedBean]}>
          <Sample />
          <StatedBeanProvider types={[TestStatedBean]}>
            <Sample />
          </StatedBeanProvider>
        </StatedBeanProvider>
      );
    };

    const app = renderer.create(<App />);
    const tree = app.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('useStatedBean and change the stated field', () => {
    const Sample = () => {
      const bean = useStatedBean(TestStatedBean);
      return (
        <div>
          <span className="field">field={bean.statedField}</span>
          <button onClick={bean.addStatedField}>add</button>
        </div>
      );
    };

    const App = () => {
      return (
        <StatedBeanProvider types={[TestStatedBean]}>
          <Sample />
        </StatedBeanProvider>
      );
    };

    const app = Enzyme.mount(<App />);

    expect(app.html().includes('field=0')).toBe(true);
    const sample = app.find('Sample');
    sample.find('button').simulate('click');
    // FIXME
    expect(app.html().includes('field=1')).toBe(false);
  });

  it('StatedBeanConsumer', () => {
    const Sample = () => {
      return (
        <StatedBeanConsumer>
          {(context: StatedBeanContextValue) => {
            expect(context).not.toBeNull();
            expect(context.container).not.toBeNull();

            return null;
          }}
        </StatedBeanConsumer>
      );
    };

    const App = () => {
      return (
        <StatedBeanProvider types={[TestStatedBean]}>
          <Sample />
        </StatedBeanProvider>
      );
    };

    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });
});
