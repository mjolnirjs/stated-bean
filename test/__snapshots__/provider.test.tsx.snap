// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react provider StatedBeanProvider 1`] = `
Array [
  <div>
    0
     
    <button
      id="addBtn"
      onClick={[Function]}
    >
      add
    </button>
  </div>,
  <div>
    0
     
    <button
      id="addBtn"
      onClick={[Function]}
    >
      add
    </button>
  </div>,
]
`;
